### PR TITLE
[HttpKernel] fix broken multiline <esi:remove>

### PR DESCRIPTION
**Pull request by sstok**
| Q | A |
| --- | --- |
| Bug Fix? | yes |
| New Feature? | n |
| BC Breaks? | n |
| Deprecations? | n |
| Tests Pass? | yes |
| Fixed Tickets |  |
| License | MIT |
| Doc PR |  |

Originally found in https://github.com/symfony/symfony/pull/14800#discussion-diff-31388942

`<esi:remove>` blocks with multiline contents were not removed.
`<esi:comment>` blocks with multiline contents were not removed.

Note. According to http://www.w3.org/TR/esi-lang
`comment is an empty element, and must not have an end tag.` so the support for multi line comments are not actually supported in the standard.

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/Tests/HttpCache/EsiTest.php [hunk 1]
```diff
@@ -92,6 +92,17 @@ public function testProcessDoesNothingIfContentTypeIsNotHtml()
         $this->assertFalse($response->headers->has('x-body-eval'));
     }
 
+    public function testRemoveEsiRemoveBlocks()
```
Comment (sstok) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/EsiTest.php]: Better naming is welcome :)

Comment (jakzal) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/EsiTest.php]: `testMultilineEsiRemoveTagsAreRemoved()`?

By the way, what about `<esi:comment>`?

Comment (sstok) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/EsiTest.php]: esi:comment doesn't accept content, it only has a text attribute.

http://www.w3.org/TR/esi-lang
`<esi:comment text="Include an ad"/>`

File: src/Symfony/Component/HttpKernel/HttpCache/Esi.php [hunk 2]
```diff
@@ -162,8 +162,9 @@ public function process(Request $request, Response $response)
 
         // we don't use a proper XML parser here as we can have ESI tags in a plain text response
         $content = $response->getContent();
-        $content = preg_replace('#<esi\:remove>.*?</esi\:remove>#', '', $content);
-        $content = preg_replace('#<esi\:comment[^>]*(?:/|</esi\:comment)>#', '', $content);
+        $content = preg_replace('#<esi\:remove>.*?</esi\:remove>#s', '', $content);
+        $content = preg_replace('#<esi\:comment[^>]+>#s', '', $content);
+        $content = preg_replace('#<esi\:comment[^>]*>(.+?)</esi\:comment>#s', '', $content);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Esi.php]: This is wrong, the specs says that esi:comment element must be empty

Comment (sstok) [inline src/Symfony/Component/HttpKernel/HttpCache/Esi.php]: Hihi I know :) but the original regex allowed this though, should we only support the standard `<comment text="some comment"/>` ?

Comment (fabpot) [inline src/Symfony/Component/HttpKernel/HttpCache/Esi.php]: I would remove anything not in the specs, that can be considered as a bug.

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/Esi.php]: To me, the original regex is fine, it does it's job

Comment (sstok) [inline src/Symfony/Component/HttpKernel/HttpCache/Esi.php]: But the original regex doesn't pass for `<esi:comment>This is removed</esi:comment>` ;) (which is is suggested by the way it looks).

Will remove the `<esi\:comment[^>]*>(.+?)</esi\:comment>` regex then :+1:

Comment (sstok) [inline src/Symfony/Component/HttpKernel/HttpCache/Esi.php]: Done

Comment (jakzal) [inline src/Symfony/Component/HttpKernel/HttpCache/Esi.php]: Aren't `<esi:comment></esi:comment>` and `<esi:comment />` equivalent?

**Comment by fabpot** [comment]
Thank you @sstok.

**Pull request closed without merging**